ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -605,6 +622,18 @@ pub trait DZKPValidator<B: UpgradableContext>: Send + Sync {
     /// when calling validate multiple times for the same base context.
     async fn validate_chunk(&self, chunk_counter: usize) -> Result<(), Error>;
 
+    /// Request and wait for validation of data associated with the supplied `record_id`.
+    ///
+    /// Validation will not generally occur immediately. The indicated record ID is marked as
+    /// complete and the future blocks pending all other records in the batch also requesting
+    /// validation. Once `validate_record` has been called for all records in the batch, the
+    /// batch is verified, and all of the `validate_record` futures complete.
+    ///
+    /// This API may only be used when the number of records per batch is the same for every
+    /// step submitting intermediates to this validator. It also requires that `set_total_records`
+    /// is set appropriately on the context that is used to create the validator.
+    async fn validate_record(&self, record_id: RecordId) -> Result<(), Error>;
```

Comment:
I've been working with MAC-based validation and I have this method on `UpgradedContext`. It works for MAC, so I am curious if you think it should work for ZKP too. Carrying both context and validator was a little annoying and didn't seem to serve the purpose or add extra safety

Reply:
That seems reasonable to me. I'm not aware of any reason it's more difficult for DZKP than for MAC. Hopefully it will also let us make validators unclonable again.